### DISCUSSION TITLE
Issue extracting change summaries of earlier version

### DISCUSSION THREAD

**Discussion by nn15296** [iTwin/itwinjs-core#1838] [Q&A] [answered]
Hi

I've been having issues when trying to use the extractChangeSummaries() method which is part of the ChangeSummaryManager class. When I specify the ChangeSummaryExtractOptions' currentVersionOnly parameter as true, it works fine. When I want to extract the change summaries from earlier versions than that of the briefcase currently cached, it throws the error pictured below. I also have the same issue with using the BriefcaseManager's processChangesets() method when trying to revert the briefcase to a previous changeset (although it works fine when advancing the briefcase to a later changeset).

![image](https://user-images.githubusercontent.com/33571096/125103040-d919cc80-e0d3-11eb-812b-2f482a7714fc.png)

I've included the code below. The first snippet involves the obtaining of the briefcaseProps. This then gets fed straight into the next code snippet which tries to extract the change summaries.

First code snippet:
    
        const localBriefcasePropsArray: LocalBriefcaseProps[] = BriefcaseManager.getCachedBriefcases(imodelId);
        if (localBriefcasePropsArray.length > 0) {
          const latestBriefcaseProps = localBriefcasePropsArray[localBriefcasePropsArray.length - 1]
          return latestBriefcaseProps;
        } else {
          const ctx = await this.createContext();
          const progressBar = new SingleBar({ format: ` [${this.getTimeNow()}]  Downloading {bar} {percentage}% | ETA: {eta}s` }, Presets.shades_classic);
          let barStarted: boolean = false;
          const downloadedBriefcaseProps = await BriefcaseManager.downloadBriefcase(ctx, {
            contextId: this.config.CONTEXT_ID, iModelId: imodelId, asOf: version.toJSON(), onProgress: (loaded: number, total: number): number => {
              if (!barStarted) {
                progressBar.start(total, 0);
                barStarted = true;
              }
              progressBar.update(loaded);
              return 0;
            },
          });
          progressBar.stop();
          return downloadedBriefcaseProps;
        }

Second code snippet:

        const ctx = await this.createContext();
        const briefcase = await BriefcaseDb.open(ctx, { fileName: briefcaseProps.fileName, readonly: false });
        const changeIds = await ChangeSummaryManager.extractChangeSummaries(ctx, briefcase);

Thanks for your help

**Comment by pmconne** [comment]
@nn15296 would you mind supplying the version of the imodeljs-backend package you are using?
@ramanujam-raman

**Reply by nn15296** [reply]
@pmconne  I'm using 2.17.2 , which I believe is the latest?

**Reply by calebmshafer** [reply]
@ramanujam-raman @diegopinate any ideas?

**Reply by diegopinate** [reply]
iModelJs is deprecating reversing changesets, so you may have better luck going from older to newer versions. I will also raise it with @ramanujam-raman that the change summary code needs to avoid reversing changesets like is doing now.

**Comment by ramanujam-raman-zz** [comment, marked as answer]
@nn15296 , ChangeSummaryManager.extractChangeSummaries() relied on reversing change sets starting from the latest version supplied. However, we have since outlawed reversing schema change sets, and the method could fail if we did encounter such a case. 

Therefore  starting with version 2.18.0, we ended up deprecating ChangeSummaryManager.extractChangeSummaries() and are replacing it with ChangeSummaryManager.createChangeSummaries() (for a range of change sets), and ChangeSummaryManger.createChangeSummary() (for a single change set - i.e., current version). 

Please consider upgrading to 2.18.0, and switching to these methods. Here are the change notes - https://www.itwinjs.org/changehistory/2.18.0/#change-summary-api-changes
